packages/code-studio/vite.config.ts

```diff
@@ -0,0 +1,123 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { defineConfig, loadEnv } from 'vite';
+import react from '@vitejs/plugin-react';
+import path from 'path';
+import dns from 'dns';
+
+// Open on localhost instead of 127.0.0.1 for Node < 17
+// https://github.com/vitejs/vite/issues/9195
+dns.setDefaultResultOrder('verbatim');
+
+// https://vitejs.dev/config/
+export default defineConfig(({ mode }) => {
+  const env = loadEnv(mode, process.cwd(), '');
+
+  // https://github.com/vitejs/vite/issues/3105#issuecomment-939703781
+  const htmlPlugin = () => ({
+    name: 'html-transform',
+    transformIndexHtml: {
+      enforce: 'pre' as const,
+      transform(html: string) {
+        return html.replace(/#(.*?)#/g, (_, p1) => env[p1]);
+      },
+    },
+  });
+
+  const packagesDir = path.resolve(__dirname, '..');
+
+  // These are paths which should be proxied to the core server
+  // https://vitejs.dev/config/server-options.html#server-proxy
+  const proxy = {};
+  const proxyPaths = [
+    env.VITE_CORE_API_URL,
+    env.VITE_NOTEBOOKS_URL,
+    env.VITE_LAYOUTS_URL,
+  ];
+
+  if (env.VITE_PROXY_URL) {
+    proxyPaths.forEach(p => {
+      proxy[p] = {
+        target: env.VITE_PROXY_URL,
+        changeOrigin: true,
+      };
+    });
+  }
+
+  return {
+    root: 'src',
+    publicDir: path.resolve(__dirname, 'public'),
+    envDir: path.resolve(__dirname),
+    server: {
+      port: Number.parseInt(env.PORT, 10) ?? 4000,
```

Comment:
The optional chaining doesn't make sense here, `Number.parseInt` always returns a `number` (`NaN` if it can't parse it).
Make a method that parses it, then returns `4000` if the parsed number is `<= 0` or `NaN`

Reply:
Good catch